File: EcoFashionBackEnd/EcoFashionBackEnd/Services/ApplicationService.cs
using AutoMapper;
using EcoFashionBackEnd.Common.Payloads.Requests;
using EcoFashionBackEnd.Dtos;
using EcoFashionBackEnd.Entities;
using EcoFashionBackEnd.Repositories;
using EcoFashionBackEnd.Services;
using Microsoft.EntityFrameworkCore;
using System.Threading.Tasks;
using System.Text.Json;

namespace EcoFashionBackEnd.Services
{
    public class ApplicationService
    {
    private readonly IRepository<Application, int> _applicationRepository;
    private readonly IMapper _mapper;
    private readonly UserService _userService;
    private readonly AppDbContext _dbContext;
    private readonly IRepository<Supplier, Guid> _supplierRepository; 
    private readonly IRepository<Designer, Guid> _designerRepository;
    private readonly CloudService _cloudService;

        public ApplicationService(
        IRepository<Application, int> applicationRepository,
        IMapper mapper,
        UserService userService,
        AppDbContext dbContext,
        IRepository<Supplier, Guid> supplierRepository, 
        IRepository<Designer, Guid> designerRepository,
        CloudService cloudService) 
    {
        _applicationRepository = applicationRepository;
        _mapper = mapper;
        _userService = userService;
        _dbContext = dbContext;
        _supplierRepository = supplierRepository;
        _designerRepository = designerRepository;
        _cloudService = cloudService;
        }

        /// <summary>
        /// Upload single image file and return URL
        /// </summary>
        private async Task<string?> UploadSingleImageAsync(IFormFile? imageFile)
        {
            if (imageFile == null || imageFile.Length <= 0)
                return null;

            try
            {
                var uploadResult = await _cloudService.UploadImageAsync(imageFile);
                return uploadResult?.SecureUrl?.ToString();
            }
            catch (Exception)
            {
                // Log error but don't fail the entire application process
                return null; // Return null instead of throwing
            }
        }

        
        /// Upload multiple portfolio files and return JSON array of URLs
        private async Task<string?> UploadPortfolioFilesAsync(List<IFormFile>? portfolioFiles)
{
    if (portfolioFiles == null || portfolioFiles.Count == 0)
        return null;

    var uploadedUrls = new List<string>();

    foreach (var file in portfolioFiles)
    {
        if (file == null || file.Length == 0)
            continue;

        try
        {
            var uploadResult = await _cloudService.UploadImageAsync(file);
            if (!string.IsNullOrWhiteSpace(uploadResult?.SecureUrl?.ToString()))
            {
                uploadedUrls.Add(uploadResult.SecureUrl.ToString());
            }
        }
        catch (Exception)
        {
            // Ghi log nếu cần, hoặc dùng ILogger
            Console.WriteLine($"❌ Upload failed for file '{file.FileName}'");
            // Optional: continue; // đã mặc định
        }
    }

    return uploadedUrls.Count > 0
        ? JsonSerializer.Serialize(uploadedUrls)
        : null;
}
 
        private async Task<ApplicationModel> MapApplicationModelAsync(Application application)
        {
            var model = _mapper.Map<ApplicationModel>(application);
            if (application.ProcessedBy.HasValue)
            {
                model.ProcessedByUser = await _userService.GetUserById(application.ProcessedBy.Value);
            }
            return model;
        }

        public async Task<ApplicationModel?> ApplyAsSupplier(int userId, ApplySupplierRequest request, IFormFile? identificationPictureFile)
        {
            // Upload profile images
            var avatarUrl = await UploadSingleImageAsync(request.AvatarFile);
            var bannerUrl = await UploadSingleImageAsync(request.BannerFile);
            var portfolioFilesJson = await UploadPortfolioFilesAsync(request.PortfolioFiles);

            var application = new Application
            {
                UserId = userId,
                TargetRoleId = 3,
                AvatarUrl = avatarUrl,
                PortfolioUrl = request.PortfolioUrl,
                PortfolioFiles = portfolioFilesJson,
                BannerUrl = bannerUrl ?? request.BannerUrl,
                SpecializationUrl = request.SpecializationUrl,
                SocialLinks = request.SocialLinks,
                IdentificationNumber = request.IdentificationNumber,
                IdentificationPicture = null,
                IsIdentificationVerified = false,
                Note = request.Note,
                PhoneNumber = request.PhoneNumber,
                Address = request.Address,
                TaxNumber = request.TaxNumber,
                IdentificationPictureOwner = request.IdentificationPictureOwner,
                Certificates = request.Certificates
            };

            if (identificationPictureFile != null && identificationPictureFile.Length > 0)
            {
                try
                {
                    var uploadResult = await _cloudService.UploadImageAsync(identificationPictureFile);
                    if (uploadResult?.SecureUrl != null)
                    {
                        application.IdentificationPicture = uploadResult.SecureUrl.ToString();
                    }
                }
                catch (Exception)
                {
                    // Log error but don't fail the application
                }
            }

            await _applicationRepository.AddAsync(application);
            var result = await _dbContext.SaveChangesAsync();
            return result > 0 ? _mapper.Map<ApplicationModel>(application) : null;
        }

        public async Task<ApplicationModel?> ApplyAsDesigner(int userId, ApplyDesignerRequest request, IFormFile? identificationPictureFile)
        {
            // Upload profile images
            var avatarUrl = await UploadSingleImageAsync(request.AvatarFile);
            var bannerUrl = await UploadSingleImageAsync(request.BannerFile);
            var portfolioFilesJson = await UploadPortfolioFilesAsync(request.PortfolioFiles);

            string? Normalize(string? value) => string.IsNullOrEmpty(value) ? null : value.Trim();

            var application = new Application
            {
                UserId = userId,
                TargetRoleId = 2,
                AvatarUrl = avatarUrl,
                PortfolioUrl = request.PortfolioUrl,
                PortfolioFiles = portfolioFilesJson,
                BannerUrl = bannerUrl ?? request.BannerUrl,
                SpecializationUrl = request.SpecializationUrl,
                SocialLinks = request.SocialLinks,
                IdentificationNumber = request.IdentificationNumber,
                IdentificationPicture = null,
                IsIdentificationVerified = false,
                Note = request.Note,
                PhoneNumber = request.PhoneNumber,
                Address = request.Address,
                TaxNumber = request.TaxNumber,
                IdentificationPictureOwner = request.IdentificationPictureOwner,
                Certificates = request.Certificates
            };

            if (identificationPictureFile != null && identificationPictureFile.Length > 0)
            {
                try
                {
                    var uploadResult = await _cloudService.UploadImageAsync(identificationPictureFile);
                    if (uploadResult?.SecureUrl != null)
                    {
                        application.IdentificationPicture = uploadResult.SecureUrl.ToString();
                    }
                }
                catch (Exception)
                {
                    // Log error but don't fail the application
                }
            }

            await _applicationRepository.AddAsync(application);
            var result = await _dbContext.SaveChangesAsync();
            return result > 0 ? _mapper.Map<ApplicationModel>(application) : null;
        }

        public async Task<ApplicationModel?> GetApplicationById(int id)
        {
            var application = await _applicationRepository.GetByIdAsync(id);
            if (application == null) return null;
            return await MapApplicationModelAsync(application);
        }

        public async Task<IEnumerable<ApplicationModel>> GetAllApplications()
        {
            var applications = await _applicationRepository.GetAll()
                .Include(a => a.User)
                .Include(a => a.Role)
                .ToListAsync();
            var result = new List<ApplicationModel>();
            foreach (var app in applications)
            {
                result.Add(await MapApplicationModelAsync(app));
            }
            return result;
        }

 
    public async Task<bool> ApproveSupplierApplication(int applicationId, int adminId)
    {
        var application = await _applicationRepository.GetByIdAsync(applicationId);
        if (application == null || application.Status != ApplicationStatus.pending || application.TargetRoleId != 3)
        {
            return false;
        }

        // Lấy User entity trực tiếp từ database
        var user = await _dbContext.Users.FindAsync(application.UserId);
        if (user == null) return false;

        var supplier = new Supplier
        {
            SupplierId = Guid.NewGuid(),
            UserId = application.UserId,
            SupplierName = user.FullName,
            AvatarUrl = application.AvatarUrl,
            PortfolioUrl = application.PortfolioUrl,
            PortfolioFiles = application.PortfolioFiles,
            BannerUrl = application.BannerUrl,
            SpecializationUrl = application.SpecializationUrl,
            Bio = application.Note,
            Email = user.Email,
            PhoneNumber = application.PhoneNumber,
            Address = application.Address,
            TaxNumber = application.TaxNumber,
            IdentificationNumber = application.IdentificationNumber,
            IdentificationPicture = application.IdentificationPicture,
            IdentificationPictureOwner = application.IdentificationPictureOwner,
            Certificates = application.Certificates,
            CreatedAt = DateTime.UtcNow,
            Status = "Active"
        };

        await _supplierRepository.AddAsync(supplier);
        
        // Cập nhật vai trò người dùng
        user.RoleId = 3; // Supplier role
        _dbContext.Users.Update(user); // Cập nhật user trong database
        
        application.Status = ApplicationStatus.approved;
        application.ProcessedAt = DateTime.UtcNow;
        application.ProcessedBy = adminId;
        application.IsIdentificationVerified = true; // Mark as verified when approved
        // TODO: Send notification to user về approval
        // TODO: Send email confirmation

        return await _dbContext.SaveChangesAsync() > 0;
    }

   public async Task<bool> ApproveDesignerApplication(int applicationId, int adminId)
        {
            var application = await _applicationRepository.GetByIdAsync(applicationId);
            if (application == null || application.Status != ApplicationStatus.pending || application.TargetRoleId != 2)
            {
                return false;
            }

            // Lấy User entity trực tiếp từ database
            var user = await _dbContext.Users.FindAsync(application.UserId);
            if (user == null) return false;

            var designer = new Designer
            {
                DesignerId = Guid.NewGuid(),
                UserId = application.UserId,
                DesignerName = user.FullName,
                AvatarUrl = application.AvatarUrl,
                PortfolioUrl = application.PortfolioUrl,
                PortfolioFiles = application.PortfolioFiles,
                BannerUrl = application.BannerUrl,
                SpecializationUrl = application.SpecializationUrl,
                Bio = application.Note, // Nếu Application có trường Bio hoặc Note
                Email = user.Email,

                // Address ; TaxNumber ; Certificates ; IdentificationNumber ; IdentificationPicture ; IdentificationPictureOwner
                PhoneNumber = application.PhoneNumber,
                Address = application.Address,
                TaxNumber = application.TaxNumber,
                IdentificationPictureOwner = application.IdentificationPictureOwner,
                Certificates = application.Certificates,

                IdentificationNumber = application.IdentificationNumber,
                IdentificationPicture = application.IdentificationPicture,

                // Certificates = null, // Nếu Application có trường này thì lấy, không thì null
                CreatedAt = DateTime.UtcNow,
                Status = "Active"
            };

            await _designerRepository.AddAsync(designer);

            // Cập nhật vai trò người dùng
            user.RoleId = 2; // Designer role
            _dbContext.Users.Update(user); // Cập nhật user trong database

            application.Status = ApplicationStatus.approved;
            application.ProcessedAt = DateTime.UtcNow;
            application.ProcessedBy = adminId;
            application.IsIdentificationVerified = true; // Mark as verified when approved
                                                         // TODO: Send notification to user về approval
                                                         // TODO: Send email confirmation

            return await _dbContext.SaveChangesAsync() > 0;
        }
        public async Task<bool> RejectApplication(int applicationId, int adminId, string rejectionReason)
        {
            var application = await _applicationRepository.GetByIdAsync(applicationId);
            if (application == null || application.Status != ApplicationStatus.pending)
            {
                return false;
            }

            application.Status = ApplicationStatus.rejected;
            application.ProcessedAt = DateTime.UtcNow;
            application.ProcessedBy = adminId;
            application.RejectionReason = rejectionReason;
            // TODO: Send notification to user về rejection với lý do
            // TODO: Send email notification

            return await _dbContext.SaveChangesAsync() > 0;
        }

    


    public async Task<IEnumerable<ApplicationModel>> FilterApplications(ApplicationStatus? status, int? targetRoleId, DateTime? createdFrom, DateTime? createdTo)
    {
        var query = _dbContext.Set<Application>()
            .Include(a => a.User)
            .Include(a => a.Role)
            .AsQueryable(); // Thêm AsQueryable() để đảm bảo kiểu

        if (status.HasValue)
        {
            query = query.Where(a => a.Status == status.Value);
        }

        if (targetRoleId.HasValue)
        {
            query = query.Where(a => a.TargetRoleId == targetRoleId.Value);
        }

        if (createdFrom.HasValue)
        {
            query = query.Where(a => a.CreatedAt >= createdFrom.Value);
        }

        if (createdTo.HasValue)
        {
            query = query.Where(a => a.CreatedAt <= createdTo.Value);
        }

        var applications = await query.ToListAsync();
        return _mapper.Map<List<ApplicationModel>>(applications);
    }

    public async Task<IEnumerable<ApplicationModel>> SearchApplications(string? keyword)
    {
        var query = _dbContext.Set<Application>()
            .Include(a => a.User)
            .Include(a => a.Role)
            .AsQueryable();

        if (!string.IsNullOrEmpty(keyword))
        {
            var lowerKeyword = keyword.ToLower();
            if (int.TryParse(keyword, out int userId))
            {
                query = query.Where(a => a.UserId == userId);
            }
            else
            {
            query = query.Where(a =>
                (!string.IsNullOrEmpty(a.IdentificationNumber) && a.IdentificationNumber.ToLower().Contains(lowerKeyword)) ||
                (!string.IsNullOrEmpty(a.Note) && a.Note.ToLower().Contains(lowerKeyword)) ||
                (a.User != null && (
                    (!string.IsNullOrEmpty(a.User.FullName) && a.User.FullName.ToLower().Contains(lowerKeyword)) || 
                    (!string.IsNullOrEmpty(a.User.Email) && a.User.Email.ToLower().Contains(lowerKeyword)) || 
                    (!string.IsNullOrEmpty(a.User.Username) && a.User.Username.ToLower().Contains(lowerKeyword))
                ))
            );
            }
        }

        var applications = await query.ToListAsync();
        return _mapper.Map<List<ApplicationModel>>(applications);
    }

    public async Task<IEnumerable<ApplicationModel>> GetApplicationsByUserId(int userId)
    {
        var applications = await _dbContext.Set<Application>()
            .Include(a => a.User)
            .Include(a => a.Role)
            .Where(a => a.UserId == userId)
            .OrderByDescending(a => a.CreatedAt)
            .ToListAsync();

        return _mapper.Map<IEnumerable<ApplicationModel>>(applications);
    }

}
}

File: EcoFashionFrontEnd/ecofashion-front-end/src/services/api/applicationService.ts
import apiClient, { handleApiResponse } from "./baseApi";

export interface ApplyDesignerRequest {
  // Profile images for landing page
  avatarFile?: File; // Ảnh đại diện
  bannerFile?: File; // Ảnh banner/cover

  portfolioUrl?: string;
  portfolioFiles?: File[]; // Multiple portfolio files
  bannerUrl?: string; // URL trực tiếp (nếu không upload file)
  specializationUrl?: string;
  certificates?: string; // CSV or JSON string
  taxNumber?: string;
  address?: string;
  identificationPictureOwner?: string;

  // Social media links as JSON string
  socialLinks?: string; // {"instagram": "url", "behance": "url", "facebook": "url"}

  identificationNumber: string;
  identificationPictureFile?: File; // File upload thay vì URL
  note?: string;
}

export interface ApplySupplierRequest {
  // Profile images for landing page
  avatarFile?: File; // Ảnh đại diện
  bannerFile?: File; // Ảnh banner/cover

  portfolioUrl?: string;
  portfolioFiles?: File[]; // Multiple portfolio files
  bannerUrl?: string; // URL trực tiếp (nếu không upload file)
  specializationUrl?: string;
  certificates?: string; // CSV or JSON string
  taxNumber?: string;
  address?: string;
  identificationPictureOwner?: string;

  // Social media links as JSON string
  socialLinks?: string; // {"instagram": "url", "behance": "url", "facebook": "url"}

  identificationNumber: string;
  identificationPictureFile?: File; // File upload thay vì URL
  note?: string;
}

export interface ApplicationModel {
  applicationId: number;
  userId: number;
  targetRoleId: number;

  // Portfolio & Profile Images
  avatarUrl?: string;
  portfolioUrl?: string;
  portfolioFiles?: string; // JSON array of file urls
  bannerUrl?: string;
  specializationUrl?: string;
  certificates?: string;
  taxNumber?: string;
  address?: string;
  identificationPictureOwner?: string;

  // Social Media
  socialLinks?: string; // JSON object of social media links

  // Identification / Xác minh
  identificationNumber?: string;
  identificationPicture?: string;
  isIdentificationVerified?: boolean;

  //Tracking
  createdAt: string;
  processedAt?: string;

  // Kết quả xử lý
  processedBy?: number;
  rejectionReason?: string;
  note?: string;

  status: "pending" | "approved" | "rejected";

  // Navigation properties
  user?: {
    userId: number;
    fullName?: string;
    email?: string;
  };
  role?: {
    roleId: number;
    roleName: string;
  };
  processedByUser?: {
    userId: number;
    fullName?: string;
    email?: string;
  };
}

export interface ApiResult<T> {
  success: boolean; // Backend sử dụng camelCase
  result: T; // Backend sử dụng camelCase
  errorMessage?: string; // Backend sử dụng camelCase
}

class ApplicationService {
  private readonly API_BASE = "Applications";

  /**
   * Helper method to create FormData for file upload
   */
  private createFormData(
    request: ApplyDesignerRequest | ApplySupplierRequest
  ): FormData {
    const formData = new FormData();

    // Profile images
    if (request.avatarFile) {
      formData.append("AvatarFile", request.avatarFile);
    }
    if (request.bannerFile) {
      formData.append("BannerFile", request.bannerFile);
    }

    // Portfolio files (multiple)
    if (request.portfolioFiles && request.portfolioFiles.length > 0) {
      request.portfolioFiles.forEach((file) => {
        formData.append("PortfolioFiles", file);
      });
    }

    // Identification file
    if (request.identificationPictureFile) {
      formData.append(
        "IdentificationPictureFile",
        request.identificationPictureFile
      );
    }

    // Text fields
    if (request.portfolioUrl)
      formData.append("PortfolioUrl", request.portfolioUrl);
    if (request.bannerUrl) formData.append("BannerUrl", request.bannerUrl);
    if (request.specializationUrl)
      formData.append("SpecializationUrl", request.specializationUrl);
    if (request.socialLinks)
      formData.append("SocialLinks", request.socialLinks);
    if (request.identificationNumber)
      formData.append("IdentificationNumber", request.identificationNumber);
    if (request.note) formData.append("Note", request.note);

    return formData;
  }

  // User đăng ký làm Designer
  async applyAsDesigner(
    request: ApplyDesignerRequest
  ): Promise<ApplicationModel> {
    const formData = this.createFormData(request);

    const response = await apiClient.post<any>(
      `${this.API_BASE}/ApplyDesigner`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
        timeout: 1200000, // 20 minutes for file upload
      }
    );
    return handleApiResponse<ApplicationModel>(response);
  }

  // User đăng ký làm Supplier
  async applyAsSupplier(
    request: ApplySupplierRequest
  ): Promise<ApplicationModel> {
    const formData = this.createFormData(request);

    const response = await apiClient.post<any>(
      `${this.API_BASE}/ApplySupplier`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
        timeout: 1200000, // 20 minutes for file upload
      }
    );
    return handleApiResponse<ApplicationModel>(response);
  }

  // User xem đơn đăng ký của mình
  async getMyApplications(): Promise<ApplicationModel[]> {
    const response = await apiClient.get<any>(
      `${this.API_BASE}/MyApplications`
    );
    return handleApiResponse<ApplicationModel[]>(response);
  }

  // Xem chi tiết đơn đăng ký
  async getApplicationById(id: number): Promise<ApplicationModel> {
    const response = await apiClient.get<any>(`${this.API_BASE}/${id}`);
    return handleApiResponse<ApplicationModel>(response);
  }

  // Admin - Xem tất cả đơn đăng ký
  async getAllApplications(): Promise<ApplicationModel[]> {
    const response = await apiClient.get<any>(`${this.API_BASE}`);
    return handleApiResponse<ApplicationModel[]>(response);
  }

  // Admin - Lọc đơn đăng ký
  async filterApplications(params: {
    status?: string;
    targetRoleId?: number;
    createdFrom?: string;
    createdTo?: string;
  }): Promise<ApplicationModel[]> {
    const queryParams = new URLSearchParams();

    if (params.status) queryParams.append("status", params.status);
    if (params.targetRoleId)
      queryParams.append("targetRoleId", params.targetRoleId.toString());
    if (params.createdFrom)
      queryParams.append("createdFrom", params.createdFrom);
    if (params.createdTo) queryParams.append("createdTo", params.createdTo);

    const response = await apiClient.get<any>(
      `${this.API_BASE}/Filter?${queryParams.toString()}`
    );

    const result = handleApiResponse<{ applications: ApplicationModel[] }>(
      response
    );
    return result.applications;
  }

  // Admin - Tìm kiếm đơn đăng ký
  async searchApplications(keyword?: string): Promise<ApplicationModel[]> {
    const queryParams = new URLSearchParams();
    if (keyword) queryParams.append("keyword", keyword);

    const response = await apiClient.get<any>(
      `${this.API_BASE}/Search?${queryParams.toString()}`
    );

    const result = handleApiResponse<{ applications: ApplicationModel[] }>(
      response
    );
    return result.applications;
  }

  // Admin - Phê duyệt Designer
  async approveDesignerApplication(applicationId: number): Promise<void> {
    const response = await apiClient.put<any>(
      `${this.API_BASE}/${applicationId}/ApproveDesigner`
    );

    handleApiResponse<string>(response);
  }

  // Admin - Phê duyệt Supplier
  async approveSupplierApplication(applicationId: number): Promise<void> {
    const response = await apiClient.put<any>(
      `${this.API_BASE}/${applicationId}/ApproveSupplier`
    );

    handleApiResponse<string>(response);
  }

  // Admin - Từ chối đơn đăng ký
  async rejectApplication(
    applicationId: number,
    rejectionReason: string
  ): Promise<void> {
    const response = await apiClient.put<any>(
      `${this.API_BASE}/${applicationId}/Reject`,
      { rejectionReason }
    );

    handleApiResponse<string>(response);
  }
}

export const applicationService = new ApplicationService();
